### ISSUE TITLE
Feature Proposal: vertical-scroll

### ISSUE THREAD

**Issue by ehsan-karamad** [WICG/feature-policy#139]
An `<iframe>` which is taking up much of the visual viewport could block users ability to scroll the main page through various ways. The propsoed feature would ensure that any `<iframe>` which does not bear the feature will not be able to block _vertical scrolling_. For context please visit the tracking [bug](https://crbug.com/611982).

Some methods which can be used for blocking (vertical)  scrolling:
1. Canceling certain events such as `touchstart`.
2. Setting `touch-action` for some element(s) which does not include `pan-y`.
3. Using scripted and programmatic scrolling such as `element.scrollIntoView()`.

Techniques (1) and (2) are easier to manage. For (1), we can mark the `TouchEvent` for specific `EventTarget` as non-cancelable. For (2) the style for any element could be overwritten to include `touch-action: pan-y` whenever the element is inside a frame which is not have the `vertical-scroll` feature.

Some of the questions that need to be addressed:
1. How to enforce the feature? Current WIP [CL](https://chromium-review.googlesource.com/c/chromium/src/+/930328) assumes the feature is enabled for `self` and any frame which does not have the feature in its policy cannot block scrolling.
2. We could consider the portion of visual viewport covered by the <iframe> as factor, i.e., if the `<iframe>` covers more than `90%` of the visual viewport _and_ is _not_ entitled to blocking `vertical-scroll` then make sure it cannot block it. 
3. What to do with programmatic scrolling? One approach could be to
  a. Do not propagate recursive scrolling to parent frame if current frame is now allowed `vertical scroll`.
  b. Do not allow programmatic scrolling altogether.
Naturally 3.a seems much better than 3.b.

**Comment by clelland** [comment]
I think that we might have to make the default allowlist for this one '`*`', to be compatible with the existing web. Using `self` means that no cross origin frame will be granted `vertical-scroll` without an explicit policy to say otherwise.

Using `*` means that you can explicitly *disable* it, with something like `<iframe src="..." allow="vertical-scroll none">`, and that will disable it for that frame and all nested content.

It might be possible to couple this with an intervention (maybe mobile-only) that turns this on for sites which have frames that cover more than 90% of the viewport, but I don't think we can make that the default.

**Comment by clelland** [comment]
Is programmatic scrolling even an issue? I would think that it's fine for such a framed document to use `scrollIntoView()`, in response to events, as long as it's not stopping a user gesture from scrolling the parent page.

**Comment by ehsan-karamad** [comment]
I agree with `'*'` being more  web-compatible. Alternatively we could make this a _negative_ feature per se. For instance `'enforce-vertical-scroll'` where by default it is disabled for all. But when enabled for a given frame, then all cross-origin frame nested inside the given frame are not allowed to block vertical scroll. Although applying the feature through sandbox seems to have better granularity here.

**Comment by ehsan-karamad** [comment]
Programmatic scrolling could be an issue if a frame constantly uses `scrollIntoView()`. This would effectively block vertical or any form of scrolling. But the problem goes beyond blocking vertical scrolling in that, even if we suppress all such calls during user scroll, after the scroll gesture ends the frame can again `scrollIntoView()` and void any changes to scroll position by the user. I also think that `scrollIntoView()` can be regarded as a separate issue which somewhat overlaps with "blocking vertical scrolling". 

My current suggestion is to limit the scope of `scrollIntoView()` to its own frame as in any element will only become visible inside the frame itself but the frame (when not allowed to) will not be scrolling in its parent. Another (perhaps less interesting) alternative could be to only change the "x" value of scroll offset in parent frames (for scrolls initiated inside a sandboxed iframe).

**Comment by clelland** [comment]
> My current suggestion is to limit the scope of `scrollIntoView()` to its own frame as in any element will only become visible inside the frame itself but the frame (when not allowed to) will not be scrolling in its parent.

Got it -- that makes sense, and that's actually what I thought `scrollIntoView()` was already doing.

**Comment by clelland** [comment]
There's an explainer document for this now in the repo at [policies/vertical_scroll.md](https://github.com/WICG/feature-policy/blob/gh-pages/policies/vertical_scroll.md)

**Comment by Malvoz** [comment]
> There's an explainer document for this now in the repo at [policies/vertical_scroll.md](https://github.com/WICG/feature-policy/blob/gh-pages/policies/vertical_scroll.md)

For anyone intending to follow that link, it has moved to: 
https://github.com/WICG/feature-policy/blob/master/policies/vertical_scroll.md
